minidump-processor/src/stackwalker/amd64.rs

```diff
@@ -331,15 +321,36 @@ where
                 raw: MinidumpRawContext::Amd64(caller_ctx),
                 valid: MinidumpContextValidity::Some(valid),
             };
-            let mut frame = StackFrame::from_context(context, FrameTrust::Scan);
-            adjust_instruction(&mut frame, caller_ip);
-            return Some(frame);
+            return Some(StackFrame::from_context(context, FrameTrust::Scan));
         }
     }
 
     None
 }
 
+/// The most strict validation we have for instruction pointers.
+///
+/// This is only used for stack-scanning, because it's explicitly
+/// trying to distinguish between total garbage and correct values.
+/// cfi and frame_pointer approaches do not use this validation
+/// because by default they're working with plausible/trustworthy
+/// data.
+///
+/// Specifically, not using this validation allows cfi/fp methods
+/// to unwind through frames we don't have mapped modules for (such as
+/// OS APIs). This may seem confusing since we obviously don't have cfi
+/// for unmapped modules!
+///
+/// The way this works is that we will use cfi to unwind some frame we
+/// know about and *end up* in a function we know nothing about, but with
+/// all the right register values. At this point, frame pointers will
+/// often do the correct thing even though we don't know what code we're
+/// in -- until we get back into code we do know about and cfi kicks back in.
+/// At worst, this sets scanning up in a better position for success!
+///
+/// If we applied this more rigorous validation to cfi/fp methods, we
+/// would just discard the correct register values from the known frame
+/// and immediately start doing unreliable scans.
```

Comment:
The problem is if you *don't* reject things from scans for things like "doesn't map to a module" then your scan will accept basically any value! On x64 you can appeal to canonicity, and I assume you can appeal to alignment(?), but still the scan would be *way* too willing to accept anything.

This in turn would be a problem for [the refinement scans in x86 STACK WIN unwinding](https://searchfox.org/mozilla-central/rev/5c35d0f53fbd1b8b992d192cb5239bf2cc3e6e6e/toolkit/crashreporter/google-breakpad/src/processor/stackwalker_x86.cc#427-434), because you would ~always throw away the CFI result in favour of some scanned garbage.

I think the concept has potential but basically you'd need "levels" of rigorousness to tune how aggressive/permissive your validation is to the precise use-case. 

If you're bottoming out in scanning after cfi/fp completely failed, it *might* make sense to have a really permissive scan, although you'll get *tons* of scan-hallucination, which might be worse than just truncating the stack (since it confuses the user). If you're doing a "refining" scan you want to be really strict. If you're checking cfi/fp you want to be really permissive.

InstructionAddressSeemsValid can't really be a monolith.

Also god breakpad code is so hard to read properly. I think only now after looking at this code 10 times has my brain finally properly read the line `symbolizer_result != StackFrameSymbolizer::kNoError`. Just always read it as `if result != <garbage prefix>Error { return true (valid) }` which broke my brain.